fuse/nodefs/fsconnector.go

```diff
@@ -126,9 +135,23 @@ func (c *FileSystemConnector) forgetUpdate(nodeID uint64, forgetCount int) {
 
 	if forgotten, handled := c.inodeMap.Forget(nodeID, forgetCount); forgotten {
 		node := (*Inode)(unsafe.Pointer(handled))
+		if node.IsDir() && len(node.Children()) > 0 {
+			// This directory node has children that have not yet been forgotten.
+			// We don't want to remove it from the tree (i.e. remove it from its parent)
+			// because the children may have parent pointers to it.
+			// A later LOOKUP will return this node with a new NodeID.
+			return
+		}
 		node.mount.treeLock.Lock()
-		c.recursiveConsiderDropInode(node)
+		// Remove Inode from all parents
```

Comment:
Yes, but you can make the tree be in perfect sync if the FUSE daemon is the
thing acting on the file system, right?

On Wed, Mar 23, 2016 at 9:44 AM rfjakob <notifications@github.com> wrote:

> In fuse/nodefs/fsconnector.go
> <https://github.com/hanwen/go-fuse/pull/103#discussion_r57124825>:
>
> >  		node.mount.treeLock.Lock()
> > -		c.recursiveConsiderDropInode(node)
> > +		// Remove Inode from all parents
>
> That's true and it's a nice feature. The problem for pathfs is this:
> https://github.com/hanwen/go-fuse/blob/master/fuse/nodefs/fsops.go#L72 .
> Nodefs decides which node to use for a lookup, and does so based just on
> the name. This is not good enough for hard link tracking UNLESS the tree in
> nodefs is perfectly in sync with the underlying file system.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/hanwen/go-fuse/pull/103/files/004bb8a47f407fea15ba2caedcb216371ca806f4#r57124825>
>